flink-connectors/flink-orc/src/main/java/org/apache/flink/orc/OrcUtils.java

```diff
@@ -0,0 +1,2657 @@
+	}
+
+	private static void readNonNullListColumn(Object[] rows, int fieldIdx, ListColumnVector list, TypeDescription schema, long[] lengthVector) {
+
+		TypeDescription fieldType = schema.getChildren().get(0);
+		if (list.isRepeating) {
+			// TODO: <Need Feedback> but I think don't need in case of struct, list and map.
+			// need to remove this condition
+		} else {
+			if (lengthVector == null) {
+				long[] lengthVectorNested = list.lengths;
+				readField(rows, fieldIdx, fieldType, list.child, lengthVectorNested);
+			}
+			else { // list in a list
+
+				Object[] nestedList = new Object[rows.length];
+				for (int i = 0; i < rows.length; i++) {
+					nestedList[i] = new Object();
+				}
+
+				// length vector for nested list
+				long[] lengthVectorNested = list.lengths;
+
+				// read nested list
+				readField(nestedList, -1, fieldType, list.child, lengthVectorNested);
+
+				// fill outer list with nested list
+				int offset = 0;
+				int length;
+				Object[] temp;
+				if (fieldIdx == -1) { // set list of list as an object
+					for (int i = 0; i < rows.length; i++) {
+						length = (int) lengthVector[i];
+						if (length == 0) {
+							rows[i] = new Object[0];
+						}
+						else {
+							temp = new Object[length];
+							System.arraycopy(nestedList, offset, temp, 0, length);
+							offset = offset + length;
+							rows[i] = temp;
+						}
+					}
+				}
+				else { // set list of list as an field on Row
+					for (int i = 0; i < rows.length; i++) {
+						length = (int) lengthVector[i];
+						if (length == 0) {
+							((Row) rows[i]).setField(fieldIdx, new Object[0]);
+						}
+						else {
+							temp = new Object[length];
+							System.arraycopy(nestedList, offset, temp, 0, length);
+							offset = offset + length;
+							((Row) rows[i]).setField(fieldIdx, temp);
+						}
+					}
+				}
+			}
+		}
+
+	}
+
+	private static void readNonNullMapColumn(Object[] rows, int fieldIdx, MapColumnVector map, TypeDescription schema, long[] lengthVector) {
+
+		List<TypeDescription> fieldType = schema.getChildren();
+		TypeDescription keyType = fieldType.get(0);
+		TypeDescription valueType = fieldType.get(1);
+
+		if (map.isRepeating) {
+			// TODO: <Need Feedback> but I think don't need in case of struct, list and map.
+			// need to remove this condition
+		} else {
+
+			ColumnVector keys = map.keys;
+			ColumnVector values = map.values;
+			Object[] keyRows = new Object[rows.length];
+			Object[] valueRows = new Object[rows.length];
+			for (int i = 0; i < rows.length; i++) {
+				keyRows[i] = new Object();
+				valueRows[i] = new Object();
+			}
+
+			// read map kes and values
+			readField(keyRows, -1, keyType, keys, null);
+			readField(valueRows, -1, valueType, values, null);
+
+			if (lengthVector == null) {
+				long[] lengthVectorMap = map.lengths;
+				int offset = 0;
+				if (fieldIdx == -1) {
+					for (int i = 0; i < rows.length; i++) {
+						HashMap<Object, Object> resultMap = new HashMap<Object, Object>();
```

Comment:
make `private final static` method that creates the `HashMap`. The method is likely to be inlined.